### ISSUE TITLE
sdxl_train_network.py lr_te1/lr_te2 not supported?

### ISSUE THREAD

**Issue by DanPli** [kohya-ss/sd-scripts#930]
Hi there,
are the new text encoder arguments to train specific TE's not supported for sdxl_train_network? I am getting sdxl_train_network.py: error: unrecognized arguments: --train_text_encoder --learning_rate_te1=1e-9 --learning_rate_te2=1e-9

If this is the case, will you add this functionality for network training?

Thanks! :)

**Comment by antis0007** [comment]
@DanPli @kohya-ss I just got this implemented in my own installation, and 0 changes needed to be made to sdxl_train_network.py. However, I am using the bmaltais/kohya_ss GUI, and I had to make a few changes to lora_gui.py as well to get it working. Overall this is a pretty easy change to make and doesn't seem to break any current functionality (single learning rate still works, and all parsing in that case is the same, text_encoder_lr is now just a list behind the scenes).
![image](https://github.com/kohya-ss/sd-scripts/assets/31860133/459a2c4c-9c92-4c46-b600-07ae44b3c717)
I dont know if I want to make a pull request since its a pretty small change and I only have the GUI version of this repo installed locally, so here are the only changes required to get this to work:

In train_network.py:
```py
...
# 学習に必要なクラスを準備する
accelerator.print("prepare optimizer, data loader etc.")

# 後方互換性を確保するよ
try: #Normal usage
    if len(args.text_encoder_lr) == 1:
        trainable_params = network.prepare_optimizer_params(float(args.text_encoder_lr[0]), args.unet_lr, args.learning_rate)
    else:
        trainable_params = network.prepare_optimizer_params(args.text_encoder_lr, args.unet_lr, args.learning_rate)

except TypeError: #Deprecated usage (omitted learning_rate)
    accelerator.print(
        "Deprecated: use prepare_optimizer_params(text_encoder_lr, unet_lr, learning_rate) instead of prepare_optimizer_params(text_encoder_lr, unet_lr)"
    )
    if len(args.text_encoder_lr) == 1:
        trainable_params = network.prepare_optimizer_params(float(args.text_encoder_lr[0]), args.unet_lr)
    else:
        trainable_params = network.prepare_optimizer_params(args.text_encoder_lr, args.unet_lr)

optimizer_name, optimizer_args, optimizer = train_util.get_optimizer(args, trainable_params)
...
#in setup_parser, change the type to a list:
parser.add_argument("--text_encoder_lr", type=list, default=None, help="learning rate for Text Encoder(s) / Text Encoder(s)の学習率")
...
```
In lora.py:
```py
# 二つのText Encoderに別々の学習率を設定できるようにするといいかも
def prepare_optimizer_params(self, text_encoder_lr, unet_lr, default_lr):
    self.requires_grad_(True)
    all_params = []

    def enumerate_params(loras):
        params = []
        for lora in loras:
            params.extend(lora.parameters())
        return params

    if self.text_encoder_loras:
        param_data = {"params": enumerate_params(self.text_encoder_loras)}
        #NEW CODE:
        if len(text_encoder_lr) > 1: #multiple TE Lrs provided
            assert len(param_data["params"])==len(text_encoder_lr)
            for i in range(0, len(text_encoder_lr)):
                if text_encoder_lr[i] is not None:
                    param_data["lr"] = text_encoder_lr[i]
                all_params.append(param_data)
        #ORIGINAL PROCESSING:
        else:
            if text_encoder_lr is not None:
                param_data["lr"] = text_encoder_lr
            all_params.append(param_data)
```
This does mean that users of the script would need to update old configs so that floats that get input for --text_encoder_lr are given in a list [lr_1,lr_2], however by making some small changes to the GUI, this is done implicitly and hidden from basic users if they only wish to input a single lr. What do you think?

**Comment by DarkAlchy** [comment]
> Unfortunately, sdxl_train_network.py (LoRA training script) doesn't support the individual learning rates for TE1/2 yet. I will add options later.

Sure you will after the OT guys finish and you can just grab it from them.  Do what you need to do so we get this.

**Comment by antis0007** [comment]
I ran into a few roadblocks while implementing this @DarkAlchy. If you'd like to help me out with this it'd be awesome! I got to the point where feeding in the multiple TE values works, however, optimizers like Prodigy are implemented in to only take a single learning rate, and there is no easy way in kohya to set up multiple optimizers/custom training setups for a single training task (without a large rewrite that disrupts other core functionality). At this point I'm thinking about writing a few of my own SDXL-focused training scripts. Also, im not sure exactly who the "OT guys" are but I'm just trying to help out here. This is something I've wanted for a long time and I've only gotten frustrated while trying to implement it on my own. So far the best solution I could think of so far is taking the average of the provided learning rates (instead of just picking the first learning rate and discarding the rest, as the script does currently)
